src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -735,7 +617,16 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                     // DecryptData will decrypt in-place and modify these to point to the actual decrypted data, which may be smaller.
                     _decryptedBytesOffset = _internalOffset;
                     _decryptedBytesCount = payloadBytes;
-                    SecurityStatusPal status = DecryptData();
+
+                    SecurityStatusPal status;
+                    lock (_handshakeLock)
+                    {
+                        status = DecryptData();
+                        if (status.ErrorCode == SecurityStatusPalErrorCode.Renegotiate)
+                        {
+                            _handshakeWaiter = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
What happens if below we end up not calling ReplyOnReAuthenticationAsync, e.g. because AllowRenegotation is false?  That will leave _handshakeWaiter as non-null, but no one will ever complete it.  Is that going to cause problems?  Can we ensure that if this is set, it'll always be completed/removed?

Reply:
I missed that. Since throw, it may not matter but it is cleaner to deal with it explicitly. fixed.